### DISCUSSION TITLE
Legacy level metadata support ?

### DISCUSSION THREAD

**Discussion by PaulCauchois** [chrislo27/PolyrhythmMania#40] [Ideas]
I know "support" isn't really the right word since you can play the levels, but it feels like it should be a lot easier to fill the metadata of levels made in v1.0.

### Idea : Metadata editor & Automatic filling.
A separate editor just for metadata. It saves you the hassle from having to load each level individually just to change the metadata. For legacy levels, there are no required fields, those that aren't filled will just keep having "N/A" instead. And, separately from that metadata editor, when a game recognizes that a legacy level is being loaded, it fills in the known info : Duration, tempo, # of inputs, and average inputs/min.

### Justification : 
Right now you have to go to the editor, open your file, change the metadata, export it, move on to the next one. And god forbid you don't add the name of the author, the song name and the artist, in which case you have to either write something random, or go googling.
It's not that bad when you only have one or two levels, but when you're trying to update your entire library, and 80% of the levels published on Youtube have been made in the legacy version, it quickly becomes troublesome. It's a bit faster if you work with the files themselves, but screwing around with the files is something that makes me kind of uneasy as a user.
I don't think it's that big of an issue, since the levels are still playable and can be recognized by their file name, but the [Legacy Level] tag is kind of frustrating to see, and sorting/filtering doesn't work.

### Potential issues : 
It's unclear when the automatic filling should take place. As you click on the level in your library ? When opening your library ? When launching the game ? I'm fairly certain filling this info for hundreds of files at once would slow the game down dramatically, so this has to be taken into account.

**Comment by chrislo27** [comment]
I think there may be a misunderstanding with how Level Metadata is supposed to function, but I understand your grievances.

First, unless you have created a lot of levels, then you shouldn't necessarily be updating these levels. The onus is on the original level creator to update their level.

Second, required fields are important for Level Metadata to function. Additionally, if certain fields were not required, then level creators would skip this information and thus make it no better than if Level Metadata hadn't existed in the first place. You would not be able to sort/filter your Library. I will make a note that if you are "writing something random" that you are only making it more difficult for you to sort and filter your own library. Remember that the responsibility of the level is the level creator itself — when they update their level, everyone benefits.

Another point that I don't think this suggestion considers is the technical side. Level Metadata is part of a different level/project format version. If you were to partially fill out Level Metadata for a legacy level, then it would not be a legacy level anymore since it would contain Level Metadata (and thus be a newer level version). Internally, other information like the BPM statistics, number of inputs, etc are done as part of the export simulation process, so this cannot be "autofilled" by the game. It has to be verified by the level creator. 

Overall, I do understand your pain points, but keep in mind that the v1.1 update has only been out for just over a month and the game is relatively new. I would expect that as levels are created over time (think in terms of years) that they would all have Level Metadata. As for existing levels, I would urge you to contact the original creator and ask them to update their level since it benefits everyone as a whole. The Level Metadata and Library system has been designed to be easy for creators and players to use, but *migration time* will still exist — there's no getting around that.

**Reply by PaulCauchois** [reply]
As much as I hate to admit it, you're right. However, I don't think even half the level creators will be bothered to update their levels, so I'd much rather do it myself. To this end I am making a Python script to automatically look up the required information, which will make updating a level much simpler.

**Reply by chrislo27** [reply]
Good luck — if you need to reference the source code, [the save protocol is outlined in this file.](https://github.com/chrislo27/PolyrhythmMania/blob/master/core/src/main/kotlin/polyrhythmmania/container/Container.kt#L251)

Please note that levels created with external tools are not officially supported, so I would recommend against distributing these levels to others.
